File: scripts/test-repair-context-wire.py
"""19.18: captured context targets must match the final selected wire contract."""
import copy, json, re, sys, unittest
from pathlib import Path
from unittest.mock import patch
ROOT=Path(__file__).resolve().parents[1]
sys.path.insert(0,str(ROOT/'api'))
from backend.ai.prompts.source_context import normalize_source_context
from backend.application.ai_translation.request_validation import build_config
from backend.ai.translation.invocation import _translate_once
from backend.ai.translation_paths.store import execution_scope
from backend.ai.clients.base import ChatResult
from backend.ai import markers
from backend.ai.provider_registry import provider_registry

class RepairContextWire(unittest.TestCase):
    def test_explicit_mapping_and_legacy(self):
        units=[{'id':'I12_P4','text':'first'},{'id':'I18_P7','text':'second'}]
        groups=[{'targetIds':['I12_P4','I18_P7','not-in-slice'], 'units':[{'id':'c','text':'captured source'}]}]
        original=copy.deepcopy(groups)
        self.assertEqual(normalize_source_context(groups, units)[0]['targetIds'], ['P0','P1'])
        self.assertEqual(normalize_source_context(groups, units, wire_ids=[u['id'] for u in units])[0]['targetIds'], ['I12_P4','I18_P7'])
        self.assertEqual(normalize_source_context(groups, units[1:], wire_ids=['I18_P7'])[0]['targetIds'], ['I18_P7'])
        self.assertEqual(groups,original)
        for ids in ([], ['I12_P4'], ['I12_P4','I12_P4'], ['', 'I18_P7']):
            with self.assertRaisesRegex(ValueError,'invalid_source_context_mapping'):
                normalize_source_context(groups, units, wire_ids=ids)

    def test_final_provider_request_main_repair_slices_and_legacy(self):
        self.assertEqual(len(provider_registry), 19)
        for spec in provider_registry:
            pid=spec.provider_id
            for conversation in (False,True):
                for selected in (['I12_P4','I18_P7'], ['I18_P7']):
                    with self.subTest(provider=pid,conversation=conversation,selected=selected):
                        units=[{'id':uid,'text':'Original source '+uid} for uid in selected]
                        payload={'translationMode':'conversation' if conversation else 'independent',
                            'units':units,'targetLang':'th','sourceLang':'en',
                            'provider':{'id':pid,'model':'fixture-context','baseUrl':spec.default_base_url,'apiKey':'fixture-key','thinking':'off'},
                            'context':{'tp_tab_session':'context-fixture-'+pid},
                            'conversation':{'documentId':'context-fixture-'+pid,'branch':'repair','origins':[
                                {'pageId':'page-'+uid.split('_')[0], 'pageOrder':int(uid.split('_')[0][1:]),
                                 'unitIds':[uid], 'originalIds':['original-'+uid]} for uid in selected]},
                            'repair':{'owner':'extension','enabled':False,'branch':'repair'},
                            'sourceContext':[{'targetIds':['I12_P4','I18_P7'], 'origin':'initial_request',
                                'units':[{'id':'c0','text':'READ ONLY EVIDENCE'}]}]}
                        cfg=build_config(payload)
                        requests=[]
                        def generate(req):
                            requests.append(req)
                            answer=json.dumps({uid:'คำแปล' for uid in req.expected_ids}) if req.response_schema else '\n'.join('<<'+(uid if uid.startswith('I') else 'TP_'+uid)+':คำแปล>>' for uid in req.expected_ids)
                            return ChatResult(text=answer,used_model='fixture-context',input_tokens=20,output_tokens=8,total_tokens=28,finish_reason='stop',terminal_completed=True,terminal_evidence='provider_done')
                        source=markers.apply([row['text'] for row in units])
                        with patch.object(spec.adapter,'generate',side_effect=generate), patch('backend.ai.provider_resolution.discovered_model_capabilities',return_value=(False,{})):
                            with execution_scope(cfg,'th'):
                                _translate_once(source,'th',cfg,is_retry=True)
                        self.assertEqual(len(requests),1)
                        req=requests[0];user='\n'.join(req.user_parts)
                        targets=[uid for group in re.findall(r'"appliesTo":(\[[^\]]*\])',user) for uid in json.loads(group)]
                        self.assertTrue(targets,user)
                        self.assertEqual(targets,list(req.expected_ids))
                        self.assertEqual(list(req.expected_ids),selected if conversation else [f'P{i}' for i in range(len(units))])

if __name__=='__main__':unittest.main(verbosity=2)

File: scripts/test-repair-context-wire.mjs
import assert from 'node:assert/strict';
import { normalizeSourceContext } from '../src/shared/ai/source-context.js';
import { composeCanonicalPrompt, composeTranslationUserMessage } from '../src/shared/ai/direct-local/prompt.js';
import { BUNDLED_CANONICAL_PROMPT_PLANS } from '../src/generated/canonical-prompt-plans.js';
const all=[{id:'I12_P4',text:'Hello'},{id:'I18_P7',text:'Goodbye'}];
const groups=[{targetIds:all.map(u=>u.id),origin:'initial_request',units:[{id:'c0',text:'Source context'}]}];
const original=structuredClone(groups);let checks=0;
for(const units of [all,all.slice(1)])for(const conversation of [true,false])for(const lang of ['th','en','ja']){
 const wire=units.map((u,i)=>conversation?u.id:`P${i}`);
 const normalized=normalizeSourceContext(groups,units,wire);
 assert.deepEqual(normalized[0].targetIds,wire);checks++;
 const prompt=composeCanonicalPrompt(BUNDLED_CANONICAL_PROMPT_PLANS[lang],{source_context:groups,style_examples:false},false,false,lang,units,wire);
 const user=composeTranslationUserMessage({sections:prompt.sections,sourceRecords:wire.map(id=>`<<${id.startsWith('I')?id:'TP_'+id}:source>>`).join('\n'),targetLang:lang,expectedIds:wire,conversationRecords:conversation});
 const targets=[...user.matchAll(/"appliesTo":(\[[^\]]*\])/g)].flatMap(m=>JSON.parse(m[1]));
 assert.deepEqual(targets,wire);checks++;
}
assert.deepEqual(normalizeSourceContext(groups,all)[0].targetIds,['P0','P1']);checks++;
for(const wire of [[],['I12_P4'],['I12_P4','I12_P4'],['','I18_P7']]){assert.throws(()=>normalizeSourceContext(groups,all,wire),/invalid_source_context_mapping/);checks++;}
assert.deepEqual(groups,original);checks++;
console.log(`PASS ${checks} source-context mapping and final prompt checks: sparse IDs, slices, legacy, th/en/ja`);

// Capture the actual fetch payload after Direct Local generation and real
// conversation preparation. Preserve existing committed messages byte-for-byte.
const { translateWithLocalOpenAi } = await import('../src/shared/ai/direct-local/generation.js');
const { localProviderCatalog, localAiPreset } = await import('../src/shared/ai/providers/local-registry.js');
const { prepareConversation, appendPreparedMessages } = await import('../src/shared/ai/conversation/prompt.js');
const oldFetch = globalThis.fetch;
let payloadCases = 0;
try {
 for (const spec of localProviderCatalog()) for (const units of [all,all.slice(1)]) {
  const ids=units.map(u=>u.id),calls=[];
  const history=[{anchor:true,user:'ANCHOR_BYTES\n<<I1_P0:OLD_SOURCE>>',assistant:'<<I1_P0:คำเก่า>>'}];
  const frozenHistory=structuredClone(history);
  const ai={provider:spec.id,model:'wire-fixture',base_url:spec.baseUrl,local_adapter:localAiPreset(spec.id),
    prompt:'',translation_mode:'conversation',style_examples:false,thinking:'off',source_context:groups,
    model_capabilities:{limits:{contextTokens:65536},structuredOutput:{supported:false},reasoning:{supported:false,control:'none'}},
    conversation:{branch:'repair',origins:ids.map(id=>({pageId:'page-'+id.split('_')[0],pageOrder:Number(id.split('_')[0].slice(1)),unitIds:[id],originalIds:['original-'+id]}))}};
  globalThis.fetch=async (url,init)=>{
   const body=JSON.parse(init.body);calls.push(body);
   const content=ids.map(id=>`<<${id}:คำแปล>>`).join('\n');
   const data=spec.protocol==='ollama'?{message:{content},done:true,done_reason:'stop',prompt_eval_count:20,eval_count:8}:
    {choices:[{message:{content},finish_reason:'stop'}],usage:{prompt_tokens:20,completion_tokens:8,total_tokens:28}};
   return new Response(JSON.stringify(data),{headers:{'content-type':'application/json'}});
  };
  await translateWithLocalOpenAi(units,{ai,canonicalPrompt:BUNDLED_CANONICAL_PROMPT_PLANS.th,targetLang:'th',sourceLang:'en',
    conversationContext:{prepare:input=>prepareConversation({...input,ai,state:{history,revision:1,scope:'wire-fixture',storage:'memory'}}),
     messages:appendPreparedMessages,capture(){}}});
  assert.equal(calls.length,1);
  assert.deepEqual(calls[0].messages.slice(1,3),[
    {role:'user',content:history[0].user},{role:'assistant',content:history[0].assistant}]);
  const user=calls[0].messages.at(-1).content;
  const targets=[...user.matchAll(/"appliesTo":(\[[^\]]*\])/g)].flatMap(m=>JSON.parse(m[1]));
  assert.deepEqual(targets,ids,spec.id+' provider-visible context');
  assert.deepEqual(history,frozenHistory,'must not rewrite committed history');
  assert.equal(calls[0].format,undefined);assert.equal(calls[0].response_format,undefined);
  payloadCases++;
 }
} finally {globalThis.fetch=oldFetch;}
console.log(`PASS ${payloadCases} captured Direct Local repair payloads across ${localProviderCatalog().length} runtimes; exact history prefix + sliced stable context IDs + one call`);
